pgx/src/memcxt.rs

```diff
@@ -369,15 +414,39 @@ impl PgMemoryContexts {
     ///
     /// ```rust,no_run
     /// use pgx::PgMemoryContexts;
-    /// let copy = PgMemoryContexts::CurrentMemoryContext.pstrdup("make a copy of this");
+    /// // SAFETY:  `CurrentMemoryContext` is always valid, and we promise not to use `copy` after
+    /// // `CurrentMemoryContext` has been free'd.
+    /// let copy = unsafe { PgMemoryContexts::CurrentMemoryContext.pstrdup("make a copy of this") };
     /// ```
-    pub fn pstrdup(&self, s: &str) -> *mut std::os::raw::c_char {
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because we cannot ensure that any of the [`PgMemoryContexts`] variants,
+    /// specifically those with payloads, actually represent valid Postgres [`pg_sys::MemoryContextData`]
+    /// pointers.
+    ///
+    /// We also cannot ensure that the result of this function will stay allocated as long as Rust's
+    /// borrow checker thinks it will.
+    pub unsafe fn pstrdup(&self, s: &str) -> *mut std::os::raw::c_char {
         let cstring = std::ffi::CString::new(s).unwrap();
         unsafe { pg_sys::MemoryContextStrdup(self.value(), cstring.as_ptr()) }
     }
 
     /// Copies `len` bytes, starting at `src` into this memory context and
     /// returns a raw `*mut T` pointer to the newly allocated location
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if `src` is a null pointer.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because we cannot ensure that any of the [`PgMemoryContexts`] variants,
+    /// specifically those with payloads, actually represent valid Postgres [`pg_sys::MemoryContextData`]
+    /// pointers.
+    ///
+    /// We also cannot ensure that the result of this function will stay allocated as long as Rust's
+    /// borrow checker thinks it will.
```

Comment:
It's not really the borrow checker that's relevant here (since the raw pointer probably will have no lifetime), but I don't have a better suggestion for wording, and this gets the point across

Reply:
The returned thing is an owned `R`, that presumably would have been allocated in the memory context since the user called `.switch_to()`.

So not really any kind of "raw pointer" is necessarily involved here (unless that's what the returned `R` actually is).